cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/schedule/trigger/TriggeringInfoHelper.java

```diff
@@ -0,0 +1,197 @@
+      case TIME:
+        DefaultTimeTriggerInfo timeTrigger = (DefaultTimeTriggerInfo) triggerInfo;
+        return new TriggeringInfo.TimeTriggeringInfo(
+          scheduleId, null, timeTrigger.getCronExpression());
+      case PARTITION:
+        DefaultPartitionTriggerInfo partTrigger = (DefaultPartitionTriggerInfo) triggerInfo;
+        return new TriggeringInfo.PartitionTriggeringInfo(
+          scheduleId, null, partTrigger.getDatasetName(), partTrigger.getDatasetNamespace(),
+          partTrigger.getExpectedNumPartitions(), partTrigger.getActualNumPartitions());
+      default:
+        return null;
+    }
+  }
+
+  /**
+   * Serializer/Deserializer for {@link TriggeringInfo}
+   */
+  public static class TriggeringInfoCodec implements JsonDeserializer<TriggeringInfo>,
+    JsonSerializer<TriggeringInfo> {
+
+    private static final Map<Trigger.Type, Class<? extends TriggeringInfo>> TYPE_TO_TRIGGER_INFO =
+      generateMap();
+
+    private static Map<Trigger.Type, Class<? extends TriggeringInfo>> generateMap() {
+      Map<Trigger.Type, Class<? extends TriggeringInfo>> map = new HashMap<>();
+      map.put(Trigger.Type.AND, TriggeringInfo.AndTriggeringInfo.class);
+      map.put(Trigger.Type.OR, TriggeringInfo.OrTriggeringInfo.class);
+      map.put(Trigger.Type.PROGRAM_STATUS, TriggeringInfo.ProgramStatusTriggeringInfo.class);
+      map.put(Trigger.Type.TIME, TriggeringInfo.TimeTriggeringInfo.class);
+      map.put(Trigger.Type.PARTITION, TriggeringInfo.PartitionTriggeringInfo.class);
+      return map;
+    }
+
+    private final Map<Trigger.Type, Class<? extends TriggeringInfo>> typeClassMap;
+
+    public TriggeringInfoCodec() {
+      this(TYPE_TO_TRIGGER_INFO);
+    }
+
+    protected TriggeringInfoCodec(Map<Trigger.Type, Class<? extends TriggeringInfo>> typeClassMap) {
+      this.typeClassMap = typeClassMap;
+    }
+
+    @Override
+    public TriggeringInfo deserialize(JsonElement json, Type type,
+                                      JsonDeserializationContext context) throws JsonParseException {
+      if (json == null) {
+        return null;
+      }
+      if (!(json instanceof JsonObject)) {
+        throw new JsonParseException("Expected a JsonObject but found a " + json.getClass().getName());
+      }
+      JsonObject object = (JsonObject) json;
+      JsonElement typeJson = object.get("type");
+      Trigger.Type triggerType = context.deserialize(typeJson, Trigger.Type.class);
+      Class<? extends TriggeringInfo> subClass = typeClassMap.get(triggerType);
+      if (subClass == null) {
+        throw new JsonParseException("Unable to map trigger type " + triggerType + " to a TriggerInfo class");
+      }
+      return context.deserialize(json, subClass);
+    }
+
+    @Override
+    public JsonElement serialize(TriggeringInfo src, Type type, JsonSerializationContext context) {
+      return context.serialize(src, src.getClass());
```

Comment:
This doesn't seem to correspond to the deserialize method logic? Isn't that we need to serialize a type field?